plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php

```diff
@@ -145,8 +145,17 @@ private function obtain_lock_on_sku_for_concurrent_requests( $product ) {
 			return boolval( $locked );
 		}
 
-		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
-		$result = $wpdb->query( $query );
+		// The insert query can potentially result in a deadlock if there is high concurrency
+		// when trying to insert products, which will result in a false negative for SKU lock
+		// and incorrectly products not being created.
+		// To mitigate this, we will retry the query 3 times before giving up.
+		for ( $attempts = 0; $attempts < 3; $attempts++ ) {
+			// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+			$result = $wpdb->query( $query );
+			if ( false !== $result ) {
+				break;
+			}
```

Comment:
If `$result === false` should we log something?

Reply:
... And/or should we do so if `$result === false && $attemps === 2` (ie, if we are at the end of our allowed retries)?

Reply:
Good point!
What do you think about adding something like this after the loop:
```php
if ($result === false) {
	wc_get_logger()->warning(
		sprintf(
			'Failed to obtain SKU lock for product: ID "%d" with SKU "%s" after %d attempts.',
			$product_id,
			$sku,
			$attempts,
		),
		array(
			'error' => $wpdb->last_error,
		)
	);
}
```

Reply:
Perfect!